docs/design/core/waiters/README.md

```diff
@@ -0,0 +1,447 @@
+you handle the task of waiting for the table to become available.
+
+## Proposed APIs
+
+The SDK 2.x will support both sync and async waiters for service clients that have waiter-eligible operations. It will also provide a generic `Waiter` class
+which makes it possible for customers to customize polling function, define expected sucess, failure and retry conditions as well as configurations such as `maxAttempts`. 
+
+### Usage Examples
+
+#### Example 1: Using sync waiters
+
+```Java
+DynamoDbClient client = DynamoDbClient.create();
+
+DescribeTableResponse response = client.waiter().waitUtilTableExists(b -> b.tableName("table"));
+```
+
+#### Example 2: Using async waiters
+
+```Java
+DynamoDbAsyncClient asyncClient = DynamoDbAsyncClient.create();
+
+CompletableFuture<DescribeTableResponse> responseFuture = 
+    asyncClient.waiter().waitUtilTableExists(b -> b.tableName("table"));
+```
+
+*FAQ Below: "Why not create waiter operations directly on the client?"*
+
+#### Example 3: Using the generic waiter
+
+```Java
+Waiter<DescribeTableResponse> waiter =
+   Waiter.<DescribeTableResponse>builder()
+        .addAcceptor(WaiterAcceptor.successAcceptor(r -> r.table().tableStatus().equals(TableStatus.ACTIVE)))
+        .addAcceptor(WaiterAcceptor.retryAcceptor(t -> t instanceof ResourceNotFoundException))
+        .addAcceptor(WaiterAcceptor.errorAcceptor(t -> t instanceof InternalServerErrorException))
+        .maxAttempts(20)
+        .backoffStrategy(BackoffStrategy.defaultStrategy())
+        .build();
+
+// run synchronousely 
+DescribeTableResponse response = waiter.run(() -> client.describeTable(describeTableRequest));
+
+// run asychronousely
+CompletableFuture<DescribeTableResponse> responseFuture =
+      waiter.runAsync(() -> asyncClient.describeTable(describeTableRequest));
+```
+
+### `{Service}Waiter` and `{Service}AsyncWaiter`
+
+Two classes will be created for each waiter-eligible service: `{Service}Waiter` and `{Service}AsyncWaiter` (e.g. `DynamoDbWaiter`, `DynamoDbAsyncWaiter`). 
+This follows the naming strategy established by the current `{Service}Client` and `{Service}Utilities` classes.
+
+#### Example
+
+```Java
+/**
+ * Waiter utility class that waits for a resource to transition to the desired state.
+ */
+@SdkPublicApi
+@Generated("software.amazon.awssdk:codegen")
+public interface DynamoDbWaiter {
+
+    /**
+     * Poller method that waits for the table status to transition to <code>ACTIVE</code> by
+     * invoking {@link DynamoDbClient#describeTable}. It returns when the resource enters into a desired state or
+     * it is determined that the resource will never enter into the desired state.
+     *
+     * @param describeTableRequest Represents the input of a <code>DescribeTable</code> operation.
+     * @return {@link DescribeTableResponse}
+     */
+    default DescribeTableResponse waitUtilTableExists(DescribeTableRequest describeTableRequest) {
```

Comment:
Until*?

Reply:
Oops! I blame intellij for not telling me.